File: src/leetCode/java/Solution56.java
package leetCode.java;

import java.util.*;

//Intervals O(nlogn) O(n), sort需要O(nlogn) O(n)
class Solution56 {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);//依照左界排序
        List<int[]> merge = new ArrayList<>();
        for(int[] interval : intervals){
            //用last指向上一個新增的interval
            int[] last = merge.isEmpty() ? new int[]{0, 0} : merge.get(merge.size() - 1);
            if(merge.isEmpty() || last[1] < interval[0]){
                //新增新的區段
                merge.add(interval);
            }else{
                //更新已有的區段的右側
                last[1] = Math.max(last[1], interval[1]);
            }
        }

        int[][] res = new int[merge.size()][2];
        for(int i=0; i<merge.size(); i++)
            res[i] = merge.get(i);

        return res;
    }
}
/*
可能有的情況
------
 ----
	-----
*/

File: src/leetCode/c++/Solution56.cpp
//Intervals O(nlogn) O(n), sort需要O(nlogn) O(n)
class Solution56 {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        //依照左界小排到大
        if (intervals.empty()) return {};
        sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b){
                return a[0] < b[0];
            });
        vector<vector<int>> res{intervals[0]};
        for (int i = 1; i < intervals.size(); ++i) {
            if (res.back()[1] < intervals[i][0]) {
                res.push_back(intervals[i]);
            } else {
                res.back()[1] = max(res.back()[1], intervals[i][1]);
            }
        }
        return res;
    }
};